src/server/transaction/server/driver.go

```diff
@@ -0,0 +1,292 @@
+			err = fmt.Errorf("unrecognized transaction request type")
+		}
+
+		if err != nil {
+			return nil, fmt.Errorf("error running request %d of %d: %v", i+1, len(info.Requests), err)
+		}
+
+		responses = append(responses, response)
+	}
+
+	info.Responses = responses
+	return info, nil
+}
+
+func (d *driver) finishTransaction(pachClient *client.APIClient, txn *transaction.Transaction) (*transaction.TransactionInfo, error) {
+	ctx := pachClient.Ctx()
+	info := &transaction.TransactionInfo{}
+	_, err := col.NewSTM(ctx, d.etcdClient, func(stm col.STM) error {
+		err := d.transactions.ReadOnly(ctx).Get(txn.ID, info)
+		if err != nil {
+			return err
+		}
+		info, err = d.runTransaction(pachClient, stm, info)
+		if err != nil {
+			return err
+		}
+		return d.transactions.ReadWrite(stm).Delete(txn.ID)
+	})
+	if err != nil {
+		return nil, err
+	}
+	return info, nil
+}
+
+// Error to be returned when the transaction has been modified between our two STM calls
+type transactionConflictError struct{}
+
+func (e *transactionConflictError) Error() string {
+	return fmt.Sprintf("transaction could not be modified due to concurrent modifications")
+}
+
+func (d *driver) appendTransaction(
+	pachClient *client.APIClient,
+	txn *transaction.Transaction,
+	items []*transaction.TransactionRequest,
+) (*transaction.TransactionInfo, error) {
+	ctx := pachClient.Ctx()
+
+	// Run this thing in a loop in case we get a conflict, time out after some tries
+	for i := 0; i < 10; i++ {
+		// We first do a dryrun of the transaction to
+		// 1. make sure the appended request is valid
+		// 2. Capture the result of the request to be returned
+		var numRequests, numResponses int
+		var dryrunResponses []*transaction.TransactionResponse
+
+		info := &transaction.TransactionInfo{}
+		err := col.NewDryrunSTM(ctx, d.etcdClient, func(stm col.STM) error {
+			// Get the existing transaction and append the new requests
+			err := d.transactions.ReadWrite(stm).Get(txn.ID, info)
+			if err != nil {
+				return err
+			}
+
+			// Save the length so that we can check that nothing else modifies the
+			// transaction in the meantime
+			numRequests = len(info.Requests)
+			numResponses = len(info.Responses)
+			info.Requests = append(info.Requests, items...)
+
+			info, err = d.runTransaction(pachClient, stm, info)
+			if err != nil {
+				return err
+			}
+
+			dryrunResponses = info.Responses[numResponses:]
+			return nil
+		})
+
+		if err != nil {
+			return nil, err
+		}
+
+		info = &transaction.TransactionInfo{}
+		_, err = col.NewSTM(ctx, d.etcdClient, func(stm col.STM) error {
+			// Update the existing transaction with the new requests/responses
+			return d.transactions.ReadWrite(stm).Update(txn.ID, info, func() error {
+				if len(info.Requests) != numRequests || len(info.Responses) != numResponses {
+					// Someone else modified the transaction while we did the dry run
+					return &transactionConflictError{}
+				}
+
+				info.Requests = append(info.Requests, items...)
```

Comment:
So am I right in understanding that if someone does a big put-file we'll put the request objects (which contain the content) into etcd? If so that's going to be a pretty bad bottleneck if people try to transactionally ingress large amounts of data. I don't think we should fix this as part of this PR but we should probably document it as a potential gotcha. And long term we'll need to figure out how to store that data elsewhere, which I don't think we'll be too hard.

Reply:
`put file` wasn't actually tackled in this PR for this reason - it needs another path from the supported operations.  The current plan is to upload the files to object storage, then store the references (whatever is needed to update the tree) into the transaction row in etcd.